pm: power: npcx: add console expired mechanism.

MulinChao opened the pull request:
This CL adds support for console expired mechanism. It implements the
notification to power management module that the module for console is
in use. If the interval that module doesn't receive any input message
exceeds CONFIG_SOC_POWER_CONSOLE_EXPIRED_TIMEOUT, the power management
module is allowed to enter deep sleep mode. This mechanism gives a
window in which the users can organize console input.

By npcx_power_console_is_in_use(), the application can forbid ec enter deep sleep in customized pm_policy_next_state().

Signed-off-by: Mulin Chao <mlchao@nuvoton.com>

MulinChao committed:
pm: power: npcx: add console expired mechanism.

This CL adds support for console expired mechanism. It implements the
notification to power management module that the module for console is
in use. If the interval that module doesn't receive any input message
exceeds CONFIG_SOC_POWER_CONSOLE_EXPIRED_TIMEOUT, the power management
module is allowed to enter deep sleep mode. This mechanism gives a
window in which the users can organize console input.

Signed-off-by: Mulin Chao <mlchao@nuvoton.com>

MulinChao commented:
Sorry for the noise. This is a draft version. Still need more time to verify its stability.

MulinChao commented:
Passed the verification on volteer with [PR32873](https://github.com/zephyrproject-rtos/zephyr/pull/32873) ,[CL2721322](https://chromium-review.googlesource.com/c/chromiumos/platform/ec/+/2721322), and [CL2738500](https://chromium-review.googlesource.com/c/chromiumos/platform/ec/+/2738500).   
Ready for code review now.

keith-zephyr commented on soc/arm/nuvoton_npcx/common/soc_power.h:
```diff
@@ -0,0 +1,36 @@
+ * extends expired time by CONFIG_SOC_POWER_CONSOLE_EXPIRED_TIMEOUT.
+ */
+void npcx_power_console_is_in_use_refresh(void);
+
```
Can you also add an API for changing the console activity timeout?   See https://chromium.googlesource.com/chromiumos/platform/ec/+/cfbbbb93/chip/npcx/clock.c#474 for the expected usage.

MulinChao replied:
Done.

keith-zephyr commented on soc/arm/nuvoton_npcx/npcx7/Kconfig.soc:
```diff
@@ -33,3 +33,23 @@ config SOC_POWER_MANAGEMENT_TRACE
 	help
 	  Internal config to enable runtime power management traces.
+
+config SOC_POWER_CONSOLE_EXPIRED
+	bool "Enable support for console expired mechanism"
+	default y
+	depends on PM
+	help
+	  Enable support for console expired mechanism. It adds the notification
+	  to the power management module that the module for console is in use.
+	  If the interval that module doesn't receive any input message exceeds
+	  CONFIG_SOC_POWER_CONSOLE_EXPIRED_TIMEOUT, the power management module
+	  is allowed to enter deep sleep mode. This mechanism gives a window in
+	  which the users can organize console input.
+
+config SOC_POWER_CONSOLE_EXPIRED_TIMEOUT
+	int "Fixed amount of time to keep the console in use flag true"
+	default 15000
+	depends on SOC_POWER_CONSOLE_EXPIRED
+	help
+	  Fixed amount of time which unit is milli-second to keep the console in
+	  use flag true.
```
Can these configs get promoted to work with the general CONFIG_UART_CONSOLE?

MulinChao replied:
Do you mean that?
1. Replace the prefix "SOC_POWER_CONSOLE_" with "UART_CONSOLE_".
2. Move these two configs to [console driver's Kconfig file](https://github.com/zephyrproject-rtos/zephyr/blob/master/drivers/console/Kconfig#L47)

If so, I will modify it later.

MulinChao replied:
Done.

MulinChao committed:
console: introduce UART console input expired mechanism options.

This CL introduces two configuration options which allows a notification
to the power management module that the UART console module is in use
now. If the interval of console module doesn't receive any input message
exceeds expired timeout, the power management module is allowed to enter
sleep/deep sleep state and turn off the clock of console module.

This mechanism gives a window in which the users can organize input
message if CONFIG_PM is enabled.

Signed-off-by: Mulin Chao <mlchao@nuvoton.com>

jackrosenthal commented on drivers/serial/uart_npcx.c:
```diff
@@ -209,6 +212,12 @@ static void uart_npcx_isr(const struct device *dev)
 
+	/* Refresh console expired time if got UART Rx event */
+	if (IS_ENABLED(CONFIG_UART_CONSOLE_INPUT_EXPIRED) &&
+						uart_npcx_irq_rx_ready(dev)) {
```
line seems over-indented (should align to left-paren of if)

MulinChao replied:
Done.

jackrosenthal commented on drivers/serial/uart_npcx.c:
```diff
@@ -367,6 +386,10 @@ static int uart_npcx_init(const struct device *dev)
 #if defined(CONFIG_PM)
+	/* Initialize a miwu device input and its callback function */
+	npcx_miwu_init_dev_callback(&data->uart_rx_cb, &config->uart_rx_wui,
+						uart_npcx_rx_wk_isr, dev);
```
line is over-indented (should align to left-paren of the function call)

MulinChao replied:
Done.

sjg20 commented on drivers/console/Kconfig:
```diff
@@ -81,6 +81,27 @@ config UART_CONSOLE_MCUMGR
+	default 15000
+	depends on UART_CONSOLE_INPUT_EXPIRED
+	help
+	  Fixed amount of time which unit is milli-second to keep the UART
```
milliseconds
or ms

MulinChao replied:
Done.

sjg20 commented on drivers/serial/uart_npcx.c:
```diff
@@ -38,6 +38,9 @@ struct uart_npcx_data {
 	uart_irq_callback_user_data_t user_cb;
 	void *user_data;
 #endif
+#ifdef CONFIG_PM
```
Can this #ifdef be dropped?

MulinChao replied:
Please see the reply in line 388 below.

MulinChao replied:
Done.

sjg20 commented on drivers/serial/uart_npcx.c:
```diff
@@ -367,6 +386,10 @@ static int uart_npcx_init(const struct device *dev)
 #endif
 
 #if defined(CONFIG_PM)
```
Really would like to use IS_ENABLED() if possible?

MulinChao replied:
In this section, the driver registers a callback function, [uart_npcx_rx_wk_isr](https://github.com/zephyrproject-rtos/zephyr/pull/32776/files#diff-80314ccee38f06d27252b7ac2d69a582d3687c8b71808737fdbbb0bc31182fd9R298), in [line 391](https://github.com/zephyrproject-rtos/zephyr/pull/32776/files#diff-80314ccee38f06d27252b7ac2d69a582d3687c8b71808737fdbbb0bc31182fd9R391) for handling the related wake-up input event. If this PR replaces "#ifdef" with "IS_ENABLED()" here, we will get a linking error, "'uart_npcx_rx_wk_isr' undeclared", if CONFIG_PM is not enabled. That's why I didn't choose IS_ENABLED() here.

MulinChao replied:
Done.

keith-zephyr commented on drivers/serial/uart_npcx.c:
```diff
@@ -285,6 +294,16 @@ static int uart_npcx_err_check(const struct device *dev)
 	return err;
 }
 
+#ifdef CONFIG_PM
+static void uart_npcx_rx_wk_isr(const struct device *dev, struct npcx_wui *wui)
```
Replace #ifdef CONFIG_PM with __unused static void uart_npcx_rx_wk_isr().   Then you can also use IS_ENABLED() at line 369.

MulinChao replied:
Oh, yes. Thanks for reminding me. I will submit the PR with this change later.

MulinChao replied:
Done.
